File: src/main/java/com/atsuishio/superbwarfare/capability/energy/DynamicEnergyStorage.java
package com.atsuishio.superbwarfare.capability.energy;

import net.fabricmc.fabric.api.transfer.v1.transaction.TransactionContext;
import net.fabricmc.fabric.api.transfer.v1.transaction.base.SnapshotParticipant;
import team.reborn.energy.api.EnergyStorage;

import java.util.function.Supplier;

public class DynamicEnergyStorage extends SnapshotParticipant<Long> implements EnergyStorage {

    protected long amount;
    protected Supplier<Long> capacityGetter;
    protected Supplier<Long> maxInsertGetter;
    protected Supplier<Long> maxExtractGetter;

    public DynamicEnergyStorage(Supplier<Long> capacityGetter) {
        this(capacityGetter, capacityGetter, capacityGetter);
    }

    public DynamicEnergyStorage(Supplier<Long> capacityGetter, Supplier<Long> maxInsertGetter, Supplier<Long> maxExtractGetter) {
        this.capacityGetter = capacityGetter;
        this.maxInsertGetter = maxInsertGetter;
        this.maxExtractGetter = maxExtractGetter;
    }

    @Override
    protected Long createSnapshot() {
        return amount;
    }

    @Override
    protected void readSnapshot(Long snapshot) {
        amount = snapshot;
    }

    @Override
    public long insert(long maxAmount, TransactionContext transaction) {
        if (maxInsertGetter.get() <= 0 || !supportsInsertion()) return 0;
        long inserted = Math.min(maxInsertGetter.get(), Math.min(maxAmount, capacityGetter.get() - amount));
        if (inserted > 0) {
            updateSnapshots(transaction);
            amount += inserted;
            return inserted;
        }
        return 0;
    }

    @Override
    public long extract(long maxAmount, TransactionContext transaction) {
        if (maxExtractGetter.get() <= 0 || !supportsExtraction()) return 0;
        long extracted = Math.min(maxExtractGetter.get(), Math.min(maxAmount, amount));
        if (extracted > 0) {
            updateSnapshots(transaction);
            amount -= extracted;
            return extracted;
        }
        return 0;
    }

    @Override
    public long getAmount() {
        return amount;
    }

    @Override
    public long getCapacity() {
        return capacityGetter.get();
    }

    @Override
    public boolean supportsInsertion() {
        return maxInsertGetter.get() > 0;
    }

    @Override
    public boolean supportsExtraction() {
        return maxExtractGetter.get() > 0;
    }
}

File: src/main/kotlin/com/atsuishio/superbwarfare/capability/energy/DynamicEnergyStorage.kt
package com.atsuishio.superbwarfare.capability.energy

import com.atsuishio.superbwarfare.capability.api.EnergyStorage
import net.fabricmc.fabric.api.transfer.v1.transaction.TransactionContext
import java.util.function.Supplier

open class DynamicEnergyStorage @JvmOverloads constructor(
    protected val maxStorageGetter: Supplier<Int?>,
    protected val maxReceiveGetter: Supplier<Int?> = maxStorageGetter,
    protected val maxExtractGetter: Supplier<Int?> = maxStorageGetter
) : EnergyStorage(Int.MAX_VALUE.toLong()) {

    override fun insert(maxAmount: Long, transaction: TransactionContext): Long {
        updateProps()
        return super.insert(maxAmount, transaction)
    }

    override fun extract(maxAmount: Long, transaction: TransactionContext): Long {
        updateProps()
        return super.extract(maxAmount, transaction)
    }

    override fun supportsInsertion(): Boolean {
        updateProps()
        return super.supportsInsertion()
    }

    override fun supportsExtraction(): Boolean {
        updateProps()
        return super.supportsExtraction()
    }

    override fun getCapacity(): Long {
        updateProps()
        return super.getCapacity()
    }

    protected fun updateProps() {
        this.capacity = maxStorageGetter.get()!!.toLong()
        this.maxExtract = maxExtractGetter.get()!!.toLong()
        this.maxReceive = maxReceiveGetter.get()!!.toLong()
    }
}
